File: Prova di luglio 2019   FATTA/soluzioni ale/es5-reparti.c
// Alessandro Belfiore 975826
/* Schema dei processi produttivi rappresentato attraverso un array di lunghezza uguale al numero di processi
lavorativi, dove in posizione array[i], viene inserito un puntatore alla radice dell'albero rappresentante
l'impianto di produzione.*/

/* Il numero di lavoratori è contenuto all'interno del singolo nodo che viene così strutturato:
*/

struct node {
  int num_reparto,num_lavoratori;
  struct node *left;
  struct node *right;
};
typedef struct *Bit_node

int lavoratoriInCollaudo(Bit_node root) {
  if(root ==  NULL) {
    return 0;
  }
  if(root -> left == NULL && root -> right == NULL) {
    return root -> num_lavoratori;
  }
  return lavoratoriInCollaudo(root -> left) + lavoratoriInCollaudo(root -> right);
}

void Algoritmo(int *a, int n) {
  int sum = 0;
  for(int i = 0; i < n; i++) {
    sum = sum + lavoratoriInCollaudo(a[i]);
  }
  printf("I lavoratori dei reparti di collaudo dei diversi impianti di produzione sono :%d\n",sum);
}

File: Prova di luglio 2019   FATTA/es5-reparti.go
package main

type Bit_node struct {
	n_prog     int
	lavoratori int
	dx, sx     *Bit_node
}

type foresta struct {
	impianti []*Bit_node
}

func lavoratoriCollaudo(root *Bit_node) int {
	if root == nil {
		return 0
	}
	if root.dx == nil && root.sx == nil {
		k := root.lavoratori
		return k
	}
	return lavoratoriCollaudo(root.sx) + lavoratoriCollaudo((root.dx))
}

func totCollaudo(f []foresta) int {
	somma := 0
	for i := 0; i < len(f); i++ {
		somma += lavoratoriCollaudo(f.impianti[i])
	}
}
